llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -384,17 +386,78 @@ void CombinerHelper::applyCombineShuffleConcat(MachineInstr &MI,
   MI.eraseFromParent();
 }
 
+// Create a stream from 0 to n with a specified number of steps
+CombinerHelper::GeneratorType
+adderGenerator(const int32_t From, const int32_t To, const int32_t StepSize) {
+  int32_t Counter = From;
+  return [Counter, To, StepSize]() mutable {
+    std::optional<int32_t> OldCount = std::optional<int32_t>(Counter);
+    Counter += StepSize;
+    if (OldCount == (To + StepSize))
+      OldCount = {};
+    return OldCount;
+  };
+}
+
 bool CombinerHelper::tryCombineShuffleVector(MachineInstr &MI) {
```

Comment:
From what I understand, this isn't an NFC change. I think it would really help readability if you first introduce `GeneratorType` in an NFC change, and then in a follow-up commit add extra features/combines.

Reply:
I fixed this, but I do struggle a bit with seeing the difference between NFC and not. Do you have good resources for me to look at for that? I think that in the rest of the commits I did a good job of keeping them separate, but I will double-check them.